barretenberg/cpp/src/barretenberg/polynomials/univariate.hpp

```diff
@@ -281,13 +281,76 @@ template <class Fr, size_t domain_end, size_t domain_start = 0> class Univariate
 
         std::copy(evaluations.begin(), evaluations.end(), result.evaluations.begin());
 
+        static constexpr Fr inverse_two = Fr(2).invert();
         if constexpr (LENGTH == 2) {
             Fr delta = value_at(1) - value_at(0);
             static_assert(EXTENDED_LENGTH != 0);
-            for (size_t idx = domain_start; idx < EXTENDED_DOMAIN_END - 1; idx++) {
+            for (size_t idx = domain_end - 1; idx < EXTENDED_DOMAIN_END - 1; idx++) {
                 result.value_at(idx + 1) = result.value_at(idx) + delta;
             }
             return result;
+        } else if constexpr (LENGTH == 3) {
+            // Based off https://hackmd.io/@aztec-network/SyR45cmOq?type=view
+            Fr a = (value_at(2) + value_at(0)) * inverse_two - value_at(1);
+            Fr b = value_at(1) - a - value_at(0);
+            Fr a2 = a + a;
+            Fr a_mul = a2;
+            for (size_t i = 0; i < domain_end - 2; i++) {
+                a_mul += a2;
+            }
+            Fr extra = a_mul + a + b;
+            for (size_t idx = domain_end - 1; idx < EXTENDED_DOMAIN_END - 1; idx++) {
+                result.value_at(idx + 1) = result.value_at(idx) + extra;
+                extra += a2;
+            }
+            return result;
+        } else if constexpr (LENGTH == 4) {
+            static constexpr Fr inverse_six = Fr(6).invert();
+
+            // These formulas were found by inverting the matrix representation of the coefficients of the 4 evaluations
```

Comment:
could add more documentation here, or add it to the hackmd cited in LENGTH == 3

Reply:
I understand that you are basically recomputing monomial coefficeints by hand, but I think it's worth pointing that out explicitly in code, so somebody not from the crypto team wouldn't get confused by this code.